### PR TITLE
Some fixes and improvements

### PR DESCRIPTION
**Pull request by apaszke**
* Added support for indexing tensors with LongTensors (#363)
* Added `torch.unbind` - splits tensor into a tuple, removing a given dimension (#427)
* Renamed `set_index` to `_set_index` (#469)
* Improved CUDA detection in THPP (it's not possible to force no cuda) (#638)

### CODE REVIEW

File: torch/csrc/generic/Tensor.cpp [hunk 1]
```diff
@@ -437,6 +437,21 @@ static PyObject * THPTensor_(pynew)(PyTypeObject *type, PyObject *args, PyObject
...
+
+#ifdef THC_GENERIC_FILE
+#define THIndexTensor THCudaLongTensor
+#define THIndexTensor_(NAME) TH_CONCAT_2(THCudaLongTensor_,NAME)
+#define THPIndexTensor THCPLongTensor
+#define THPIndexTensor_Check THCPLongTensor_Check
+#else
+#define THIndexTensor THLongTensor
+#define THIndexTensor_(NAME) TH_CONCAT_2(THLongTensor_,NAME)
+#define THPIndexTensor THPLongTensor
```
Comment (colesbury) [inline torch/csrc/generic/Tensor.cpp]: Do these need to be #undef below?

**Review by colesbury — APPROVED** [review]
See comments. Otherwise LGTM

File: torch/csrc/generic/Tensor.cpp (hunk 1 above)
Comment (apaszke) [inline torch/csrc/generic/Tensor.cpp]: Ah I forgot, thanks

File: torch/lib/THPP/CMakeLists.txt [hunk 2]
```diff
@@ -36,15 +36,18 @@ IF(NOT TH_LIBRARIES)
...
-FIND_PACKAGE(CUDA 5.5)
-IF(CUDA_FOUND)
-  INCLUDE_DIRECTORIES(${CUDA_INCLUDE_DIRS})
-  INCLUDE_DIRECTORIES("${CUDA_SDK_ROOT_DIR}/common/inc")
-
-  IF(NOT THC_LIBRARIES)
-    SET(THC_LIBRARIES "THC")
-  ENDIF(NOT THC_LIBRARIES)
-  MESSAGE(STATUS "THC_LIBRARIES: ${THC_LIBRARIES}")
+IF(NO_CUDA)
```
Comment (apaszke) [inline torch/lib/THPP/CMakeLists.txt]: Hm I think I wanted to print something in there. I'll fix that.

**Comment by CDLuminate** [comment]
@apaszke Thanks!

**Pull request merged**